src/initializers/bull/bull.ts

```diff
@@ -0,0 +1,121 @@
+import { defaultsDeep, has } from 'lodash';
+import * as cron from 'node-cron';
+import { getLogger } from '../log4js';
+import requireInjected from '../../require-injected';
+const Queue = requireInjected('bull');
+const newrelic = process.env.NEW_RELIC_LICENSE_KEY ? requireInjected('newrelic') : null;
+
+export default class Bull {
+  private logger = getLogger('orka.bull');
+  private prefix: string;
+  private defaultOptions: any;
+  private redisOpts: any;
+  private queueOpts: { [x: string]: { options: any } };
+  private instances = {};
+  private metricsTask: any;
+
+  constructor(prefix, queueOpts, defaultOptions, redisOpts) {
+    this.prefix = prefix;
+    this.defaultOptions = defaultOptions;
+    this.queueOpts = queueOpts.reduce((acc, q) => {
+      acc[q.name] = q;
+      return acc;
+    }, {});
+    this.redisOpts = redisOpts;
+  }
+
+  private createQueue(name: string) {
+    const fullName = `${this.prefix}:${name}`;
+    const options = this.queueOpts[name].options;
+    this.logger.info(`Creating Queue: ${fullName}`);
+    const defaultJobOptions = defaultsDeep({}, options, this.defaultOptions);
+    const queue = new Queue(fullName, { redis: this.redisOpts, defaultJobOptions });
+    queue
+      .on('drained', () => {
+        getLogger(fullName).info(`Queue drained`);
+      })
+      .on('error', err => {
+        handleError(fullName, err);
+      })
+      .on('failed', (job, err) => {
+        handleFailure(fullName, job, err);
+      });
+    // Cache the instance
+    this.instances[name] = queue;
+  }
+
+  public getQueue(name: string) {
+    if (!has(this.queueOpts, name)) {
+      throw new Error('no such queue');
+    }
+    if (!has(this.instances, name)) {
+      this.createQueue(name);
+    }
+    return this.instances[name];
+  }
+
+  private async fetchStats(name): Promise<{ queue: string; count: number; failed: number }> {
+    const q = this.getQueue(name);
+    return {
+      queue: name,
+      count: await q.count(),
+      failed: await q.getFailedCount()
+    };
+  }
+
+  public getStats() {
+    const stats = Object.keys(this.queueOpts).map(name => this.fetchStats(name));
```

Comment:
nit: 
```suggestion
    const stats = Object.keys(this.queueOpts).map(this.fetchStats);
```

Reply:
your suggestion does not work unless you do a `this.fetchStats.bind(this)` which in turn raises a TS error because the return type of the bound function cannot be inferred anymore

Reply:
Try:

```
const stats = Object.keys(this.queueOpts).map(this.fetchStats, this);
```